zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -788,11 +788,13 @@ pub(crate) struct MulAddWordsGadget<F> {
 }
 
 impl<F: Field> MulAddWordsGadget<F> {
-    pub(crate) fn construct(cb: &mut ConstraintBuilder<F>) -> Self {
-        let a = cb.query_word();
-        let b = cb.query_word();
-        let c = cb.query_word();
-        let d = cb.query_word();
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        a: util::Word<F>,
+        b: util::Word<F>,
+        c: util::Word<F>,
+        d: util::Word<F>,
```

Comment:
I wonder: If `MulAddWordsGadget` is not allocating `a, b, c, d`; should it still assign them?  Or should a gadget just assign to cells it has allocated itself?

Reply:
I think it makes more sense to move the assignment of `a, b, c, d` out of the `MulAddWordsGadget`.

Reply:
Agree, they've been moved out.